vms/evm/subscriber_test.go

```diff
@@ -0,0 +1,130 @@
+package evm
+
+import (
+	"math/big"
+	"os"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/config"
+	"github.com/ava-labs/awm-relayer/database"
+	mock_ethclient "github.com/ava-labs/awm-relayer/vms/evm/mocks"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	"github.com/ava-labs/subnet-evm/interfaces"
+	"github.com/ava-labs/subnet-evm/x/warp"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/mock/gomock"
+)
+
+func makeSubscriberWithMockEthClient(t *testing.T) (*subscriber, *mock_ethclient.MockClient) {
+	sourceSubnet := config.SourceSubnet{
+		SubnetID:          "2TGBXcnwx5PqiXWiqxAKUaNSqDguXNh1mxnp82jui68hxJSZAx",
+		ChainID:           "S4mMqUXe7vHsGiRAma6bv3CKnyaLssyAxmQ2KvFpX1KEvfFCD",
+		VM:                config.EVM.String(),
+		APINodeHost:       "127.0.0.1",
+		APINodePort:       9650,
+		EncryptConnection: false,
+		RPCEndpoint:       "https://subnets.avax.network/mysubnet/rpc",
+	}
+
+	logger := logging.NewLogger(
+		"awm-relayer-test",
+		logging.NewWrappedCore(
+			logging.Info,
+			os.Stdout,
+			logging.JSON.ConsoleEncoder(),
+		),
+	)
+
+	subnetId, err := ids.FromString(sourceSubnet.ChainID)
+	if err != nil {
+		t.Fatalf("Failed to create subnet ID")
+	}
+
+	db, err := database.NewJSONFileStorage(logger, t.TempDir(), []ids.ID{subnetId})
+	if err != nil {
+		t.Fatalf("Failed to create JSON file storage")
+	}
+
+	mockEthClient := mock_ethclient.NewMockClient(gomock.NewController(t))
+	subscriber := NewSubscriber(logger, sourceSubnet, db)
+	subscriber.dial = func(_url string) (ethclient.Client, error) { return mockEthClient, nil }
+
+	return subscriber, mockEthClient
+}
+
+func TestProcessFromHeight(t *testing.T) {
+	testCases := []struct {
+		latest int64
+		input  int64
+	}{
+		{
+			latest: 1000,
+			input:  800,
+		},
+		{
+			latest: 1000,
+			input:  700,
+		},
+		{
+			latest: 19642,
+			input:  751,
+		},
+		{
+			latest: 96,
+			input:  41,
+		},
+	}
+
+	expectFilterLogs := func(
+		mock *mock_ethclient.MockClient,
+		fromBlock int64,
+		toBlock int64,
+	) {
+		mock.EXPECT().FilterLogs(
+			gomock.Any(),
+			interfaces.FilterQuery{
+				Topics: [][]common.Hash{
+					{warp.WarpABI.Events["SendWarpMessage"].ID},
+					{},
+					{},
+				},
+				Addresses: []common.Address{
+					warp.ContractAddress,
+				},
+				FromBlock: big.NewInt(fromBlock),
+				ToBlock:   big.NewInt(toBlock),
+			},
+		).Return([]types.Log{}, nil).Times(1)
+	}
+
+	min := func(a, b int64) int64 {
```

Comment:
Not recommending we update in this PR, but in Go 1.21 we can use the built-in `min`: https://tip.golang.org/ref/spec#Min_and_max

Reply:
I realized this while implementing, but thank you for pointing it out, because it inspired me to document this via a `TODO: delete this` type comment in the code, in https://github.com/ava-labs/awm-relayer/pull/89/commits/6618b41f81d0bc8edfda6b83c1f21b9a7856ceca